cpu/esp32: make CI happy

maribu opened the pull request:
### Contribution description

As the title says.

### Testing procedure

- The CI will run the `static-tests`, which should now pass for everything in `cpu/esp32/*`.
- The code changes should *NOT* translate to changed in generated binaries

### Issues/PRs references

None

gschorcht reviewed (commented):
@maribu Thanks for providing it. There are still some small problems where `cpp-check` is complaining. Could you correct them too?

maribu commented:
> There are still some small problems where `cpp-check` is complaining. Could you correct them too?

I fixed them as well (I think those were from uncrustify) and squashed them in.

I believe the missing braces actually resulting in wrong control flow, if I read the comment and the indent correctly. So I added a second commit changing this.

yarrick commented on cpu/esp32/esp_xtimer.c:
```diff
@@ -93,10 +93,10 @@ void _ets_to_xtimer_free(ETSTimer *timer)
 void IRAM_ATTR _ets_to_xtimer_callback (void *arg)
 {
-    struct _ets_to_xtimer* e2xt = (struct _ets_to_xtimer*)arg;
+    CHECK_PARAM(e2xt != NULL);
```
hmm, is this now checked before that variable is allocated? seems wierd.

What was the CI report?

maribu replied:
Indeed, I didn't look closely and assumed that it was `CHECK_PARAM(arg != NULL)` as the warning of `cppcheck` makes no sense otherwise.

Here is the output:

```
  	cpu/esp32/esp_xtimer.c:99: warning (nullPointerRedundantCheck): Either the condition 'e2xt!=NULL' is redundant or there is possible null pointer dereference: e2xt.
  	cpu/esp32/esp_xtimer.c:104: warning (nullPointerRedundantCheck): Either the condition 'e2xt!=NULL' is redundant or there is possible null pointer dereference: e2xt.
  	cpu/esp32/esp_xtimer.c:121: warning (nullPointerRedundantCheck): Either the condition 'e2xt!=NULL' is redundant or there is possible null pointer dereference: e2xt.
  	cpu/esp32/esp_xtimer.c:122: warning (nullPointerRedundantCheck): Either the condition 'e2xt!=NULL' is redundant or there is possible null pointer dereference: e2xt.
  	cpu/esp32/esp_xtimer.c:124: warning (nullPointerRedundantCheck): Either the condition 'e2xt!=NULL' is redundant or there is possible null pointer dereference: e2xt.
  	cpu/esp32/esp_xtimer.c:125: warning (nullPointerRedundantCheck): Either the condition 'e2xt!=NULL' is redundant or there is possible null pointer dereference: e2xt.
  	cpu/esp32/esp_xtimer.c:136: warning (nullPointerRedundantCheck): Either the condition 'e2xt!=NULL' is redundant or there is possible null pointer dereference: e2xt.
  	cpu/esp32/esp_xtimer.c:137: warning (nullPointerRedundantCheck): Either the condition 'e2xt!=NULL' is redundant or there is possible null pointer dereference: e2xt.
  	cpu/esp32/esp_xtimer.c:147: warning (nullPointerRedundantCheck): Either the condition 'e2xt!=NULL' is redundant or there is possible null pointer dereference: e2xt.
  	cpu/esp32/esp_xtimer.c:151: warning (nullPointerRedundantCheck): Either the condition 'e2xt!=NULL' is redundant or there is possible null pointer dereference: e2xt.
  	cpu/esp32/esp_xtimer.c:152: warning (nullPointerRedundantCheck): Either the condition 'e2xt!=NULL' is redundant or there is possible null pointer dereference: e2xt.
```

maribu replied:
I reverted the buggy changes and the cppcheck warnings pop back up. I don't get these locally, so it might be just due to an outdated version of `cppcheck`.

maribu commented:
I added lots of suppressions now instead

gschorcht commented on cpu/esp32/esp_xtimer.c:
```diff
@@ -93,10 +93,9 @@ void _ets_to_xtimer_free(ETSTimer *timer)
-
-    CHECK_PARAM (e2xt != NULL);
-    CHECK_PARAM (e2xt->ets_timer != NULL);
+    CHECK_PARAM(e2xt->ets_timer != NULL);
```
To be honest, the code seems less readable with this order. I used `CHECK_PARAM` here to avoid a bunch of `if` blocks with `DEBUG` and returns and to keep the code more readable.

The question is whether `CHECK_PARAM` should protect against soft or hard errors. For soft errors, the check with a return would be correct. For hard errors, an `assert` would be better.

Would cpp-check be happy with `assert`s in original order?

gschorcht commented on cpu/esp32/esp_xtimer.c:
```diff
@@ -118,10 +117,18 @@ void ets_timer_setfn(ETSTimer *ptimer, ETSTimerFunc *pfunc, void *parg)
     CHECK_PARAM(e2xt != NULL);
 
+    /* cppcheck-suppress nullPointerRedundantCheck
+     * CHECK_PARAM() sends cppcheck off rails */
```
Would an `assert` make cpp-check happy here?

maribu replied:
Let's give it a try :-)

maribu replied:
The `assert()`s do work.

maribu commented on cpu/esp32/periph/i2c_hw.c:
```diff
@@ -703,6 +703,7 @@ static void _i2c_transfer (i2c_t dev)
     #if FIFO_USED
     /* reset RX FIFO queue */
     _i2c_hw[dev].regs->fifo_conf.rx_fifo_rst = 1;
+    /* cppcheck-suppress redundantAssignment */
```
Is this obvious enough, or should I add a comment explaining that cppcheck is only mildly more aware of C syntax than a regex, and hence doesn't understand that two assignments to a variable with a `volatile` qualifier are for the side affect this causes, not because the value written is of actual interest? Maybe in a bit less snarky words?

I always have the feeling that the reasoning in cppcheck suppression are more about explaining why cppcheck is a poor tool, rather than explaining what the code does. And that the end result is less readable code and reduced maintainability ...

gschorcht replied:
Perhaps a short statement would be good that it is a desired behavior here to trigger a reset pulse.

gschorcht replied:
Please squash directly.

gschorcht reviewed (commented):
Looks good.

maribu commented on cpu/esp32/periph/can.c:
```diff
@@ -259,16 +259,26 @@ static int _esp_can_send(candev_t *candev, const struct can_frame *frame)
     esp_frame.rtr = (frame->can_id & CAN_RTR_FLAG);
     esp_frame.eff = (frame->can_id & CAN_EFF_FLAG);
 
+    /* esp_frame is a union, a concept that cppcheck does not understand. For
+     * this reason cppcheck incorrectly detects values of assignments to union
+     * members as unused (unreadVariable). But their memory is accesses via
+     * union aliases */
```
Note: I was unable to reproduce this locally. The most recent `cppcheck` does understand `union`s and is aware that dead writes to `volatile` variables are usually intentional, but it needs full access to all headers. I think here it is missing the include path to `stdint.h`. Or maybe the CI is just using an outdated version of `cppcheck`. I will adapt the comment a bit to be less snarky about `cppcheck`, as missing include paths (if this is indeed the issue) is not really the fault of the tool.

maribu commented:
I think we could undo all of the changes here once `cppcheck` gets all required all required `-I` paths. Maybe it also needs to be updated to the most recent version for all of the false positives to vanish.

I'm very confident that `cppcheck` is not the lousy tool I first assumed and once we give it a bit of love will stop issuing the stupid false positives it currently does. This leads to the question should we really get this PR in, or just see if I can manage to quickly fix the `cppcheck` scripts to provide the required include paths.

gschorcht commented:
> I think we could undo all of the changes here once `cppcheck` gets all required all required `-I` paths.

Is that something we can fix by platform implementation, or is that an issue of the make system or CI?

maribu commented:
The script is in `dist/tools/cppcheck/check.sh` and relatively trivial.

It is not totally trivial to provide a set of "canonical" include paths, as e.g. `cpu.h`, `periph_conf.h`, `periph_cpu.h`, `board.h` etc. are provided at multiple locations and depending on the C file to check, it does matter which valid combination of these are actually used.

I guess checking all C files in `drivers` for one board (or better one board per architecture) would be fine. But for C files in `cpu/`, one does not want the `cpu.h` of STM32 to be used when checking ATmega periphal drivers.

gschorcht commented:
Hm, Murdock isn't happy:
```
UserWarning:/redis-py/works/best/with/hiredis./Please/consider/installing
/bin/sh: 1: /usr/local/lib/python3.8/dist-packages/redis/connection.py:77:: not found
```

maribu commented:
thanks for the reviews!
